pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,347 @@ package certconfig
+	certOperatorID = "cert-operator"
+
+	loopbackIP = "127.0.0.1"
+
+	// systemMastersOrganization is the RBAC kubernetes admin group.
+	systemMastersOrganization = "system:masters"
+)
+
+var (
+	kubeAltNames = []string{
+		"kubernetes",
+		"kubernetes.default",
+		"kubernetes.default.svc",
+		"kubernetes.default.svc.cluster.local",
+	}
 )
 
+// GetDesiredState returns all desired CertConfigs for managed certificates.
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig, err := prepareClusterConfig(*r.baseClusterConfig, *clusterGuestConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredCertConfigs := make([]*v1alpha1.CertConfig, 0, len(managedCertificates))
+	for _, mc := range managedCertificates {
+		certConfig := mc.certConfigFactoryFunc(clusterConfig, mc.name, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+
+	return desiredCertConfigs, nil
+}
+
+func prepareClusterConfig(baseClusterConfig cluster.Config, clusterGuestConfig v1alpha1.ClusterGuestConfig) (*cluster.Config, error) {
+	var err error
+
+	// Copy baseClusterConfig as basis and supplement it with information from
+	// clusterGuestConfig.
+	clusterConfig := baseClusterConfig
+
+	clusterConfig.ClusterID = key.ClusterID(clusterGuestConfig)
+
+	clusterConfig.Domain.API, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.APICert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Calico, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.CalicoCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Etcd, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.EtcdCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.NodeOperator, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.NodeOperatorCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Prometheus, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.PrometheusCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.ServiceAccount, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.ServiceAccountCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterConfig.Domain.Worker, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.WorkerCert))
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	versionBundle, err := versionbundle.GetBundleByName(key.VersionBundles(clusterGuestConfig), certOperatorID)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig.VersionBundleVersion = versionBundle.Version
+
+	return &clusterConfig, nil
```

Comment:
Personal nitpick. I find it weird to deal with a value inside the function but then return a pointer. I think when we return a pointer we should deal with a pointer internally. Adding the ampersand at the end when we return feels too magic to me. Somebody is going to refactor this later, wants to do an early return and has to change this here or has to remember to make the value a pointer in yet another place.

Reply:
If `clusterConfig` is pointer directly, the above `key` function will then require `*` which also clutters readability a bit. I'm fine with both. Which way should we go?

Reply:
I don't have an opinion, but I don't think it will require `*`.

Reply:
Ah, you are right. I must have mixed this with `baseClusterConfig` which actually now requires separate allocation and then copy. I must have dealt with the mentioned key issue somewhere else when creating this.

I don't think this is so complicated, but not so straightforward either.